lottie/slot: support custom effect slot overriding

tinyjin opened the pull request:
This patch enables slot overriding for custom effects (LottieFxCustom).

Allowing Lottie animations to dynamically override effect properties for Custom effect
through the slot system, providing runtime customization of visual effects.


### Sample (default effect slot)

- [effect-default-slot.json](https://github.com/user-attachments/files/23117184/slot12.json)
- [multiple-effect-slot.json](https://github.com/user-attachments/files/23120664/slot12.json)

### Result (lottie-web vs ThorVG)

1. Single property test

![CleanShot 2025-10-23 at 01 38 14](https://github.com/user-attachments/assets/313b41a6-1caa-4ef1-8d64-b8b646e15bd6)

2. Multiple property test

![CleanShot 2025-10-24 at 19 05 55](https://github.com/user-attachments/assets/225266df-8afd-4ff3-ad53-04858222b1c8)


issue: https://github.com/thorvg/thorvg/issues/3826

hermet commented on src/loaders/lottie/tvgLottieModel.h:
```diff
@@ -142,9 +190,8 @@ struct LottieFxCustom : LottieEffect
     }
 
-    LottieProperty* property(const char* name)
+    LottieProperty* propertyById(unsigned long id)
```
Is it necessary to change the method name?

hermet reviewed (commented):
@tinyjin Thanks, Please check comments, only reviewed: "lottie: revise LottieEffect model to inherit from LottieObject"

tinyjin commented on src/loaders/lottie/tvgLottieModel.h:
```diff
@@ -1017,7 +1022,7 @@ struct LottieLayer : LottieGroup
     LottieEffect* effectById(unsigned long id)
     {
         ARRAY_FOREACH(p, effects) {
-            if (id == (*p)->nm || id == (*p)->mn) return *p;
```
LottieObject has `id` property that is same value as `nm` in effect. So the same value is parsed where parseCommon is fired and stores value in `obj->id`. Also I changed effect object uses `obj->id` instead of `effect->nm`.

I guess the value can be unified, that is reason I changed on this as long as `effect->nm` and `obj->id` remain the same.

<img width="1226" height="302" alt="Image" src="https://github.com/user-attachments/assets/ece92dd6-2bd2-4e32-939c-ab027e33aceb" />

<img width="1270" height="510" alt="Image" src="https://github.com/user-attachments/assets/d30a0cf1-fab5-451a-b386-b523c543a8cd" />

tinyjin committed:
lottie: enhance slot overriding for custom effect

- revise LottieEffect model to inherit from LottieObject
- compact LottieEffect model
- refactor slot backup logic to support extended cases

This patch enables slot overriding for custom effects (LottieFxCustom).

Allowing Lottie animations to dynamically override effect properties for Custom effect
through the slot system, providing runtime customization of visual effects.

issue: https://github.com/thorvg/thorvg/issues/3826

hermet reviewed (commented):
@tinyjin Reviewed "lottie: revise LottieEffect model to inherit from LottieObject". 

Please check a comment. Thanks.

hermet reviewed (commented):
Reviewed: "lottie/slot: support custom effect slot overriding"

hermet commented on src/loaders/lottie/tvgLottieModel.cpp:
```diff
@@ -206,55 +206,7 @@ void LottieSlot::apply(LottieProperty* prop, bool byDefault)
-            default: break;
-        }
-    }
+    ARRAY_FOREACH(pair, pairs) pair->prop = pair->obj->override(prop, !copy);
```
```
ARRAY_FOREACH(pair, pairs) {
    pair->prop = pair->obj->override(prop, release);
}
```

hermet commented on src/loaders/lottie/tvgLottieModel.h:
```diff
@@ -167,6 +167,51 @@ struct LottieFxCustom : LottieEffect
+                    auto cur = static_cast<LottieFloat*>(target);
+                    if (release) cur->release();
+                    else backup = new LottieFloat(*cur);
+
```
--

hermet commented on src/loaders/lottie/tvgLottieLoader.cpp:
```diff
@@ -349,7 +349,10 @@ uint32_t LottieLoader::gen(const char* slots, bool byDefault)
             if ((*p)->sid != sid) continue;  //find target
-            if (auto prop = parser.parse(*p)) custom->props.push({prop, *p});
+            if (auto prop = parser.parse(*p)) {
+              prop->sid = sid;
```
please check indentation.

hermet commented on src/loaders/lottie/tvgLottieParser.cpp:
```diff
@@ -1326,6 +1331,11 @@ bool LottieParser::parseEffect(LottieEffect* effect, void(LottieParser::*func)(L
             else skip();
         }
+
+        if (sid && property) {
```
Please consider this: 
`if (sid && property) registerSlot(effect, sid, property);`

hermet reviewed (commented):
@tinyjin please check comments. thanks.

hermet commented on src/loaders/lottie/tvgLottieLoader.cpp:
```diff
@@ -349,7 +349,10 @@ uint32_t LottieLoader::gen(const char* slots, bool byDefault)
-            if (auto prop = parser.parse(*p)) custom->props.push({prop, *p});
+            if (auto prop = parser.parse(*p)) {
+                prop->sid = sid;
+                custom->props.push({prop, *p});
```
LottieSlot already has a `sid`, maybe you can assign `sid` inside parser.parse().

hermet commented on src/loaders/lottie/tvgLottieParser.cpp:
```diff
@@ -1326,6 +1331,8 @@ bool LottieParser::parseEffect(LottieEffect* effect, void(LottieParser::*func)(L
             else if (property && KEY_AS("mn")) property->mn = djb2Encode(getString());
             else skip();
         }
+
```
--

hermet reviewed (commented):
@tinyjin ++nitpicks

tinyjin commented:
@theashraf Lemme chcek :)
May I have your test files? (OR did you test with same samples here?)
